// @felte/common/dist/esm/index.js
export { _some } from './utils/some.js';
export { _mapValues } from './utils/mapValues.js';
export { isFieldValue } from './utils/isFieldValue.js';
export { _cloneDeep } from './utils/cloneDeep.js';
export { _isPlainObject } from './utils/isPlainObject.js';
export { _mergeWith } from './utils/mergeWith.js';
export { _defaultsDeep } from './utils/defaultsDeep.js';
export { _merge } from './utils/merge.js';
export { _get } from './utils/get.js';
export { _set } from './utils/set.js';
export { _unset } from './utils/unset.js';
export { _update } from './utils/update.js';
export { deepSet } from './utils/deepSet.js';
export { deepSome } from './utils/deepSome.js';
export { getPath } from './utils/getPath.js';
export { shouldIgnore } from './utils/shouldIgnore.js';
export { getValue } from './utils/getValue.js';
export { mergeErrors, runValidations } from './utils/executeValidation.js';
export { executeTransforms } from './utils/executeTransforms.js';
export { createId } from './utils/createId.js';
export { isEqual } from './utils/isEqual.js';
export { debounce } from './utils/debounce.js';
export { isElement, isFieldSetElement, isFormControl, isInputElement, isSelectElement, isTextAreaElement } from './utils/typeGuards.js';
export { addAttrsFromFieldset, getFormControls, getFormDefaultValues, getInputTextOrNumber, setControlValue, setForm } from './utils/domUtils.js';
